README.rst
Docker Custodian
================

.. |tests badge| image:: https://github.com/yelp/docker-custodian/actions/workflows/tests.yml/badge.svg
.. |docker badge| image:: https://github.com/yelp/docker-custodian/actions/workflows/docker.yml/badge.svg
.. |pypi badge| image:: https://github.com/yelp/docker-custodian/actions/workflows/publish.yml/badge.svg

|tests badge| |docker badge| |pypi badge|

Keep docker hosts tidy.


.. contents::
    :backlinks: none

Install
-------

There are three installation options

Container
~~~~~~~~~

.. code::

    docker pull yelp/docker-custodian
    docker run -ti \
        -v /var/run/docker.sock:/var/run/docker.sock \
        yelp/docker-custodian --help

Debian/Ubuntu package
~~~~~~~~~~~~~~~~~~~~~

First build the package (requires `dh-virtualenv`)

.. code:: sh

    dpkg-buildpackage -us -uc

Then install it

.. code:: sh

    dpkg -i ../docker-custodian_*.deb


Source
~~~~~~

.. code:: sh

    pip install git+https://github.com/Yelp/docker-custodian.git#egg=docker_custodian


dcgc
----

Remove old docker containers and docker images.

``dcgc`` will remove stopped containers and unused images that are older than
"max age".  Running containers, and images which are used by a container are
never removed.

Maximum age can be specificied with any format supported by
`pytimeparse <https://github.com/wroberts/pytimeparse>`_.

Example:

.. code:: sh

    dcgc --max-container-age 3days --max-image-age 30days


Prevent images from being removed
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``dcgc`` supports an image exclude list. If you have images that you'd like
to keep around forever you can use the exclude list to prevent them from
being removed.

::

    --exclude-image
        Never remove images with this tag. May be specified more than once.

    --exclude-image-file
        Path to a file which contains a list of images to exclude, one
        image tag per line.

You also can use basic pattern matching to exclude images with generic tags.

.. code::

    user/repositoryA:*
    user/repositoryB:?.?
    user/repositoryC-*:tag


Prevent containers and associated images from being removed
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

``dcgc`` also supports a container exclude list based on labels.  If there are
stopped containers that you'd like to keep, then you can check the labels to
prevent them from being removed.

::

    --exclude-container-label
        Never remove containers that have the label key=value. =value can be
        omitted and in that case only the key is checked. May be specified
        more than once.

You also can use basic pattern matching to exclude generic labels.

.. code::

    foo*
    com.docker.compose.project=test*
    com.docker*=*bar*


dcstop
------

Stop containers that have been running for too long.

``dcstop`` will ``docker stop`` containers where the container name starts
with `--prefix` and it has been running for longer than `--max-run-time`.


Example:

.. code:: sh

    dcstop --max-run-time 2days --prefix "projectprefix_"


CHANGELOG.md
# Changelog

## [v0.7.4](https://github.com/yelp/docker-custodian/tree/v0.7.4) (2021-06-23)

[Full Changelog](https://github.com/yelp/docker-custodian/compare/v0.7.3...v0.7.4)

**Closed issues:**

- Upgrade requests to version 2.20.0 or later \(CVE-2018-18074\) [\#50](https://github.com/Yelp/docker-custodian/issues/50)
- --dangling-volumes runs into an error when there are no dangling volumes [\#41](https://github.com/Yelp/docker-custodian/issues/41)
- Use kwargs\_from\_env\(\) to get Client kwargs [\#13](https://github.com/Yelp/docker-custodian/issues/13)

**Merged pull requests:**

- Move to GitHub Actions [\#61](https://github.com/Yelp/docker-custodian/pull/61) ([IamTheFij](https://github.com/IamTheFij))
- Remove pre-commit autoupdate in tox [\#60](https://github.com/Yelp/docker-custodian/pull/60) ([IamTheFij](https://github.com/IamTheFij))
- Enable multi-arch builds [\#59](https://github.com/Yelp/docker-custodian/pull/59) ([ViViDboarder](https://github.com/ViViDboarder))
- Simplify instructions for installing with pip [\#54](https://github.com/Yelp/docker-custodian/pull/54) ([dmerejkowsky](https://github.com/dmerejkowsky))
- Bump urllib and requests [\#52](https://github.com/Yelp/docker-custodian/pull/52) ([keymone](https://github.com/keymone))

## [v0.7.3](https://github.com/yelp/docker-custodian/tree/v0.7.3) (2019-04-25)

[Full Changelog](https://github.com/yelp/docker-custodian/compare/v0.7.2...v0.7.3)

**Merged pull requests:**

- Fix null labels [\#51](https://github.com/Yelp/docker-custodian/pull/51) ([mattmb](https://github.com/mattmb))
- adding .secrets.baseline [\#49](https://github.com/Yelp/docker-custodian/pull/49) ([domanchi](https://github.com/domanchi))
- Ignore CHANGELOG.md for the end-of-file precommit hook [\#48](https://github.com/Yelp/docker-custodian/pull/48) ([ATRAN2](https://github.com/ATRAN2))

## [v0.7.2](https://github.com/yelp/docker-custodian/tree/v0.7.2) (2018-03-21)

[Full Changelog](https://github.com/yelp/docker-custodian/compare/v0.7.1...v0.7.2)

## [v0.7.1](https://github.com/yelp/docker-custodian/tree/v0.7.1) (2018-03-21)

[Full Changelog](https://github.com/yelp/docker-custodian/compare/v0.6.1...v0.7.1)

**Implemented enhancements:**

- Remove volumes when removing containers [\#36](https://github.com/Yelp/docker-custodian/pull/36) ([pauloconnor](https://github.com/pauloconnor))
- Add patterns support for exclude-image-file [\#35](https://github.com/Yelp/docker-custodian/pull/35) ([vshlapakov](https://github.com/vshlapakov))

**Fixed bugs:**

- Fix broken filtering for images used by containers [\#43](https://github.com/Yelp/docker-custodian/pull/43) ([chekunkov](https://github.com/chekunkov))

**Closed issues:**

- docker-custodian crashing when cleaning up. [\#38](https://github.com/Yelp/docker-custodian/issues/38)
- backports.ssl-match-hostname\>=3.5 error when running in Docker [\#37](https://github.com/Yelp/docker-custodian/issues/37)
- Publish on PyPI? [\#10](https://github.com/Yelp/docker-custodian/issues/10)

**Merged pull requests:**

- Add --exclude-container-label argument to dcgc [\#47](https://github.com/Yelp/docker-custodian/pull/47) ([ATRAN2](https://github.com/ATRAN2))
- Port to docker lib [\#46](https://github.com/Yelp/docker-custodian/pull/46) ([samiam](https://github.com/samiam))
- Correctly handle empty dangling volumes [\#45](https://github.com/Yelp/docker-custodian/pull/45) ([samiam](https://github.com/samiam))
- Remove dangling volumes [\#40](https://github.com/Yelp/docker-custodian/pull/40) ([ymilki](https://github.com/ymilki))
- Revert docker-py version change [\#39](https://github.com/Yelp/docker-custodian/pull/39) ([pauloconnor](https://github.com/pauloconnor))

## [v0.6.1](https://github.com/yelp/docker-custodian/tree/v0.6.1) (2016-08-31)

[Full Changelog](https://github.com/yelp/docker-custodian/compare/v0.5.1...v0.6.1)

**Closed issues:**

- Add ability to kill/delete running containers by age [\#25](https://github.com/Yelp/docker-custodian/issues/25)
- Add tag to yelp/docker-custodian on the docker hub [\#21](https://github.com/Yelp/docker-custodian/issues/21)
- Error while fetching server API version [\#20](https://github.com/Yelp/docker-custodian/issues/20)
- Could It also clear mesos tmp files? [\#19](https://github.com/Yelp/docker-custodian/issues/19)

**Merged pull requests:**

- Bump to python \>=2.7 [\#33](https://github.com/Yelp/docker-custodian/pull/33) ([danielhoherd](https://github.com/danielhoherd))
- Overriding tox with /bin/true [\#32](https://github.com/Yelp/docker-custodian/pull/32) ([danielhoherd](https://github.com/danielhoherd))
- Bump debian version to 0.6.0 [\#31](https://github.com/Yelp/docker-custodian/pull/31) ([danielhoherd](https://github.com/danielhoherd))
- Remove py26 support, code cleanup [\#27](https://github.com/Yelp/docker-custodian/pull/27) ([kentwills](https://github.com/kentwills))
- Don't remove recently created containers that were never used [\#23](https://github.com/Yelp/docker-custodian/pull/23) ([analogue](https://github.com/analogue))
- Upgrade requirements to fix \#17 [\#18](https://github.com/Yelp/docker-custodian/pull/18) ([dnephin](https://github.com/dnephin))

## [v0.5.1](https://github.com/yelp/docker-custodian/tree/v0.5.1) (2015-09-29)

[Full Changelog](https://github.com/yelp/docker-custodian/compare/v0.4.0...v0.5.1)

**Implemented enhancements:**

- Use Client\(version='auto'\) [\#7](https://github.com/Yelp/docker-custodian/issues/7)

**Fixed bugs:**

- Automatically determining API version doesn't work? [\#17](https://github.com/Yelp/docker-custodian/issues/17)

**Closed issues:**

- Put on the docker hub using the yelp namespace [\#16](https://github.com/Yelp/docker-custodian/issues/16)
- Is it production ready? [\#15](https://github.com/Yelp/docker-custodian/issues/15)

**Merged pull requests:**

- Automatically determine the Docker API version to use [\#12](https://github.com/Yelp/docker-custodian/pull/12) ([jschrantz](https://github.com/jschrantz))

tests/conftest.py
import datetime

from dateutil import tz
import docker
try:
    from unittest import mock
except ImportError:
    import mock
import pytest


@pytest.fixture
def container():
    return {
        'Id': 'abcdabcdabcdabcd',
        'Created': '2013-12-20T17:00:00Z',
        'Name': '/container_name',
        'State': {
            'Running': False,
            'FinishedAt': '2014-01-01T17:30:00Z',
            'StartedAt': '2014-01-01T17:01:00Z',
        }
    }


@pytest.fixture
def image():
    return {
        'Id': 'abcdabcdabcdabcd',
        'Created': '2014-01-20T05:00:00Z',
    }


@pytest.fixture
def now():
    return datetime.datetime(2014, 1, 20, 10, 10, tzinfo=tz.tzutc())


@pytest.fixture
def earlier_time():
    return datetime.datetime(2014, 1, 1, 0, 0, tzinfo=tz.tzutc())


@pytest.fixture
def later_time():
    return datetime.datetime(2014, 1, 20, 0, 10, tzinfo=tz.tzutc())


@pytest.fixture
def mock_client():
    client = mock.create_autospec(docker.APIClient)
    client._version = '1.21'
    return client


tests/args_test.py
import datetime
try:
    from unittest import mock
except ImportError:
    import mock

from dateutil import tz

from docker_custodian import args


def test_datetime_seconds_ago(now):
    expected = datetime.datetime(2014, 1, 15, 10, 10, tzinfo=tz.tzutc())
    with mock.patch(
        'docker_custodian.args.datetime.datetime',
        autospec=True,
    ) as mock_datetime:
        mock_datetime.now.return_value = now
        assert args.datetime_seconds_ago(24 * 60 * 60 * 5) == expected


def test_timedelta_type_none():
    assert args.timedelta_type(None) is None


def test_timedelta_type(now):
    expected = datetime.datetime(2014, 1, 15, 10, 10, tzinfo=tz.tzutc())
    with mock.patch(
        'docker_custodian.args.datetime.datetime',
        autospec=True,
    ) as mock_datetime:
        mock_datetime.now.return_value = now
        assert args.timedelta_type('5 days') == expected


docker_custodian/docker_autostop.py
#!/usr/bin/env python
"""
Stop docker container that have been running longer than the max_run_time and
match some prefix.
"""
import argparse
import logging
import sys

import dateutil.parser
import docker
import docker.errors
import requests.exceptions

from docker_custodian.args import timedelta_type
from docker.utils import kwargs_from_env


log = logging.getLogger(__name__)


def stop_containers(client, max_run_time, matcher, dry_run):
    for container_summary in client.containers():
        container = client.inspect_container(container_summary['Id'])
        name = container['Name'].lstrip('/')
        if (
            matcher(name) and
            has_been_running_since(container, max_run_time)
        ):

            log.info("Stopping container %s %s: running since %s" % (
                container['Id'][:16],
                name,
                container['State']['StartedAt']))

            if not dry_run:
                stop_container(client, container['Id'])


def stop_container(client, id):
    try:
        client.stop(id)
    except requests.exceptions.Timeout as e:
        log.warn("Failed to stop container %s: %s" % (id, e))
    except docker.errors.APIError as ae:
        log.warn("Error stopping %s: %s" % (id, ae))


def build_container_matcher(prefixes):
    def matcher(name):
        return any(name.startswith(prefix) for prefix in prefixes)
    return matcher


def has_been_running_since(container, min_time):
    started_at = container.get('State', {}).get('StartedAt')
    if not started_at:
        return False

    return dateutil.parser.parse(started_at) <= min_time


def main():
    logging.basicConfig(
        level=logging.INFO,
        format="%(message)s",
        stream=sys.stdout)

    opts = get_opts()
    client = docker.APIClient(version='auto',
                              timeout=opts.timeout,
                              **kwargs_from_env())

    matcher = build_container_matcher(opts.prefix)
    stop_containers(client, opts.max_run_time, matcher, opts.dry_run)


def get_opts(args=None):
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--max-run-time',
        type=timedelta_type,
        help="Maximum time a container is allows to run. Time may "
        "be specified in any pytimeparse supported format."
    )
    parser.add_argument(
        '--prefix', action="append", default=[],
        help="Only stop containers which match one of the "
             "prefix."
    )
    parser.add_argument(
        '--dry-run', action="store_true",
        help="Only log actions, don't stop anything."
    )
    parser.add_argument(
        '-t', '--timeout', type=int, default=60,
        help="HTTP timeout in seconds for making docker API calls."
    )
    opts = parser.parse_args(args=args)

    if not opts.prefix:
        parser.error("Running with no --prefix will match nothing.")

    return opts


if __name__ == "__main__":
    main()


docker_custodian/docker_gc.py
#!/usr/bin/env python
"""
Remove old docker containers and images that are no longer in use.

"""
import argparse
import fnmatch
import logging
import sys
import time

import dateutil.parser
import docker
import docker.errors
import requests.exceptions

from collections import namedtuple
from docker_custodian.args import timedelta_type
from docker.utils import kwargs_from_env

log = logging.getLogger(__name__)


# This seems to be something docker uses for a null/zero date
YEAR_ZERO = "0001-01-01T00:00:00Z"

ExcludeLabel = namedtuple('ExcludeLabel', ['key', 'value'])


def cleanup_containers(
    client,
    max_container_age,
    dry_run,
    exclude_container_labels,
):
    all_containers = get_all_containers(client)
    filtered_containers = filter_excluded_containers(
        all_containers,
        exclude_container_labels,
    )
    for container_summary in reversed(list(filtered_containers)):
        container = api_call(
            client.inspect_container,
            container=container_summary['Id'],
        )
        if not container or not should_remove_container(
            container,
            max_container_age,
        ):
            continue

        log.info("Removing container %s %s %s" % (
            container['Id'][:16],
            container.get('Name', '').lstrip('/'),
            container['State']['FinishedAt']))

        if not dry_run:
            api_call(
                client.remove_container,
                container=container['Id'],
                v=True,
            )


def filter_excluded_containers(containers, exclude_container_labels):
    if not exclude_container_labels:
        return containers

    def include_container(container):
        if should_exclude_container_with_labels(
            container,
            exclude_container_labels,
        ):
            return False
        return True
    return filter(include_container, containers)


def should_exclude_container_with_labels(container, exclude_container_labels):
    if container['Labels']:
        for exclude_label in exclude_container_labels:
            if exclude_label.value:
                matching_keys = fnmatch.filter(
                    container['Labels'].keys(),
                    exclude_label.key,
                )
                label_values_to_check = [
                    container['Labels'][matching_key]
                    for matching_key in matching_keys
                ]
                if fnmatch.filter(label_values_to_check, exclude_label.value):
                    return True
            else:
                if fnmatch.filter(
                    container['Labels'].keys(),
                    exclude_label.key
                ):
                    return True
    return False


def should_remove_container(container, min_date):
    state = container.get('State', {})

    if state.get('Running'):
        return False

    if state.get('Ghost'):
        return True

    # Container was created, but never started
    if state.get('FinishedAt') == YEAR_ZERO:
        created_date = dateutil.parser.parse(container['Created'])
        return created_date < min_date

    finished_date = dateutil.parser.parse(state['FinishedAt'])
    return finished_date < min_date


def get_all_containers(client):
    log.info("Getting all containers")
    containers = client.containers(all=True)
    log.info("Found %s containers", len(containers))
    return containers


def get_all_images(client):
    log.info("Getting all images")
    images = client.images()
    log.info("Found %s images", len(images))
    return images


def get_dangling_volumes(client):
    log.info("Getting dangling volumes")
    volumes = client.volumes({'dangling': True})['Volumes'] or []
    log.info("Found %s dangling volumes", len(volumes))
    return volumes


def cleanup_images(
    client,
    max_image_age,
    recent_image_age,
    dry_run,
    exclude_set
):
    # re-fetch container list so that we don't include removed containers

    containers = get_all_containers(client)
    images = get_all_images(client)
    if docker.utils.compare_version('1.21', client._version) < 0:
        image_tags_in_use = {container['Image'] for container in containers}
        images = filter_images_in_use(images, image_tags_in_use)
    else:
        # ImageID field was added in 1.21
        image_ids_in_use = {container['ImageID'] for container in containers}
        images = filter_images_in_use_by_id(images, image_ids_in_use)
    images = filter_excluded_images(images, exclude_set)

    if recent_image_age is not None:
        images = without_recently_used_images(client, images, recent_image_age)

    for image_summary in reversed(list(images)):
        remove_image(client, image_summary, max_image_age, dry_run)


def without_recently_used_images(client, images, recent_image_age):
    exclude = set()
    for event in client.events(since=recent_image_age,
                               until=int(time.time()),
                               decode=True):
        status = event.get('status', '')
        if status != 'create' and not status.startswith('exec_start'):
            continue
        image_ref = event.get('from')
        if image_ref is None:
            continue
        exclude.add(image_ref)
    return [
        image for image in images
        if exclude.intersection(
                [image.get('Id')] +
                image.get('RepoTags', []) +
                image.get('RepoDigests', [])
        )
    ]


def filter_excluded_images(images, exclude_set):
    def include_image(image_summary):
        image_tags = image_summary.get('RepoTags')
        if no_image_tags(image_tags):
            return True
        for exclude_pattern in exclude_set:
            if fnmatch.filter(image_tags, exclude_pattern):
                return False
        return True

    return filter(include_image, images)


def filter_images_in_use(images, image_tags_in_use):
    def get_tag_set(image_summary):
        image_tags = image_summary.get('RepoTags')
        if no_image_tags(image_tags):
            # The repr of the image Id used by client.containers()
            return set(['%s:latest' % image_summary['Id'][:12]])
        return set(image_tags)

    def image_not_in_use(image_summary):
        return not get_tag_set(image_summary) & image_tags_in_use

    return filter(image_not_in_use, images)


def filter_images_in_use_by_id(images, image_ids_in_use):
    def image_not_in_use(image_summary):
        return image_summary['Id'] not in image_ids_in_use

    return filter(image_not_in_use, images)


def is_image_old(image, min_date):
    return dateutil.parser.parse(image['Created']) < min_date


def no_image_tags(image_tags):
    return not image_tags or image_tags == ['<none>:<none>']


def remove_image(client, image_summary, min_date, dry_run):
    image = api_call(client.inspect_image, image=image_summary['Id'])
    if not image or not is_image_old(image, min_date):
        return

    log.info("Removing image %s" % format_image(image, image_summary))
    if dry_run:
        return

    image_tags = image_summary.get('RepoTags')
    # If there are no tags, remove the id
    if no_image_tags(image_tags):
        api_call(client.remove_image, image=image_summary['Id'])
        return

    # Remove any repository tags so we don't hit 409 Conflict
    for image_tag in image_tags:
        api_call(client.remove_image, image=image_tag)


def remove_volume(client, volume, dry_run):
    if not volume:
        return

    log.info("Removing volume %s" % volume['Name'])
    if dry_run:
        return

    api_call(client.remove_volume, name=volume['Name'])


def cleanup_volumes(client, dry_run):
    dangling_volumes = get_dangling_volumes(client)

    for volume in reversed(dangling_volumes):
        log.info("Removing dangling volume %s", volume['Name'])
        remove_volume(client, volume, dry_run)


def api_call(func, **kwargs):
    try:
        return func(**kwargs)
    except requests.exceptions.Timeout as e:
        params = ','.join('%s=%s' % item for item in kwargs.items())
        log.warn("Failed to call %s %s %s" % (func.__name__, params, e))
    except docker.errors.APIError as ae:
        params = ','.join('%s=%s' % item for item in kwargs.items())
        log.warn("Error calling %s %s %s" % (func.__name__, params, ae))


def format_image(image, image_summary):
    def get_tags():
        tags = image_summary.get('RepoTags')
        if not tags or tags == ['<none>:<none>']:
            return ''
        return ', '.join(tags)

    return "%s %s" % (image['Id'][:16], get_tags())


def build_exclude_set(image_tags, exclude_file):
    exclude_set = set(image_tags or [])

    def is_image_tag(line):
        return line and not line.startswith('#')

    if exclude_file:
        lines = [line.strip() for line in exclude_file.read().split('\n')]
        exclude_set.update(filter(is_image_tag, lines))
    return exclude_set


def format_exclude_labels(exclude_label_args):
    exclude_labels = []
    for exclude_label_arg in exclude_label_args:
        split_exclude_label = exclude_label_arg.split('=', 1)
        exclude_label_key = split_exclude_label[0]
        if len(split_exclude_label) == 2:
            exclude_label_value = split_exclude_label[1]
        else:
            exclude_label_value = None
        exclude_labels.append(
            ExcludeLabel(
                key=exclude_label_key,
                value=exclude_label_value,
            )
        )
    return exclude_labels


def main():
    logging.basicConfig(
        level=logging.INFO,
        format="%(message)s",
        stream=sys.stdout)

    args = get_args()
    client = docker.APIClient(version='auto',
                              timeout=args.timeout,
                              **kwargs_from_env())

    exclude_container_labels = format_exclude_labels(
        args.exclude_container_label
    )

    if args.max_container_age:
        cleanup_containers(
            client,
            args.max_container_age,
            args.dry_run,
            exclude_container_labels,
        )

    if args.max_image_age:
        exclude_set = build_exclude_set(
            args.exclude_image,
            args.exclude_image_file)
        cleanup_images(
            client,
            args.max_image_age,
            args.recent_image_age,
            args.dry_run,
            exclude_set
        )

    if args.dangling_volumes:
        cleanup_volumes(client, args.dry_run)


def get_args(args=None):
    parser = argparse.ArgumentParser()
    parser.add_argument(
        '--max-container-age',
        type=timedelta_type,
        help="Maximum age for a container. Containers older than this age "
             "will be removed. Age can be specified in any pytimeparse "
             "supported format.")
    parser.add_argument(
        '--max-image-age',
        type=timedelta_type,
        help="Maxium age for an image. Images older than this age will be "
             "removed. Age can be specified in any pytimeparse supported "
             "format.")
    parser.add_argument(
        '--recent-image-age',
        type=timedelta_type, default=None,
        help="Images used within specified time interval will not be removed.")
    parser.add_argument(
        '--dangling-volumes',
        action="store_true",
        help="Dangling volumes will be removed.")
    parser.add_argument(
        '--dry-run', action="store_true",
        help="Only log actions, don't remove anything.")
    parser.add_argument(
        '-t', '--timeout', type=int, default=60,
        help="HTTP timeout in seconds for making docker API calls.")
    parser.add_argument(
        '--exclude-image',
        action='append',
        help="Never remove images with this tag.")
    parser.add_argument(
        '--exclude-image-file',
        type=argparse.FileType('r'),
        help="Path to a file which contains a list of images to exclude, one "
             "image tag per line.")
    parser.add_argument(
        '--exclude-container-label',
        action='append', type=str, default=[],
        help="Never remove containers with this label key or label key=value")

    return parser.parse_args(args=args)


if __name__ == "__main__":
    main()
